src/pqm/pqm.py

```diff
@@ -450,63 +104,83 @@ def _pqm_test(
-        counts = counts_tensor.round().long().cpu().numpy()
-        Nx, Ny, Ng = counts.tolist()
-    
+    counts = np.random.multinomial(
+        num_refs,
+        [x_frac * (1.0 - gauss_frac), (1.0 - x_frac) * (1.0 - gauss_frac), gauss_frac],
+    )
+    Nx, Ny, Ng = counts
+
     # Validate counts
     if Nx + Ny + Ng != num_refs:
         raise ValueError(
             f"Something went wrong. Nx={Nx}, Ny={Ny}, Ng={Ng} should sum to num_refs={num_refs}"
         )
-    
-    # Sampling reference indices
-    if is_numpy:
-        refs = _sample_reference_indices_numpy(Nx, nx, Ny, ny, Ng, x_samples, y_samples)
-    elif is_torch:
-        refs = _sample_reference_indices_torch(Nx, nx, Ny, ny, Ng, x_samples, y_samples, device)
-    
-    # Update num_refs in case Gaussian samples were added
-    current_num_refs = Nx + Ny + Ng
-    
-    # Compute nearest references and counts
-    if is_numpy:
-        return _compute_distances_numpy(x_samples, y_samples, refs, current_num_refs, num_refs)
-    
-    elif is_torch:
-        return _compute_distances_torch(x_samples, y_samples, refs, current_num_refs, num_refs)
+    if Nx >= x_samples.shape[0] - 2:
+        raise ValueError("Cannot sample more references from x_samples than available")
+    if Ny >= y_samples.shape[0] - 2:
+        raise ValueError("Cannot sample more references from y_samples than available")
+
+    # count samples in each voronoi bin
+    if is_torch:
+        refs, x_samples, y_samples = _sample_reference_indices_torch(
+            Nx, Ny, Ng, x_samples, y_samples, device
+        )
+        counts_x, counts_y = _compute_counts_torch(x_samples, y_samples, refs, num_refs)
+    else:
+        refs, x_samples, y_samples = _sample_reference_indices_numpy(
+            Nx, Ny, Ng, x_samples, y_samples
+        )
+        counts_x, counts_y = _compute_counts_numpy(x_samples, y_samples, refs, num_refs)
+
+    # Remove references with no counts
+    C = (counts_x > 0) | (counts_y > 0)
+    counts_x = counts_x[C]
+    counts_y = counts_y[C]
+
+    n_filled_bins = C.sum().item()
+    if n_filled_bins == 1:
+        raise ValueError(
+            """
+            Only one Voronoi cell has samples, so chi^2 cannot 
+            be computed. This is likely due to a small number 
+            of samples or a pathological distribution. If possible, 
+            increase the number of x_samples and y_samples.
+            """
+        )
+    if n_filled_bins < (num_refs // 2):
+        warnings.warn(
+            """
+            Less than half of the Voronoi cells have any samples in them.
+            Possibly due to a small number of samples or a pathological
+            distribution. Result may be unreliable. If possible, increase the
+            number of x_samples and y_samples.
+            """
+        )
+
+    # Perform chi-squared test using SciPy
+    contingency_table = np.stack([counts_x, counts_y])
```

Comment:
Have you tested to make sure this is as fast/faster than the torch version?

Reply:
Yes, both run at the same speed, just under 1 milisecond